### PR TITLE
Fix CI build for Azure Pipelines, use .NET 5.0 Preview

### PR DESCRIPTION
**Pull request by couven92**
*This PR is based on the yet unmerged PR #115 using the `episode21` branch.*
*This PR closes #118 as it fully contains all changes made there.*

# TL;DR

* You should use the `UseDotNet@2` task (not the deprecated `DotNetCoreInstaller` task)
* Instead of using a particular version, use the `5.0.x` wildcard version and enable preview builds
* The .NET 5 Preview SDK that is currently listed on https://dotnet.microsoft.com/download/dotnet/5.0 unfortunately does not work correctly with the Source Generators feature yet (at least it does not work as we intend it to work)  
   **Temporarily(!)** we use the [dotnet-install scripts](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-install-script) to install .NET Core 3.1.3 Preview SDK in the way the `UseDotNet@2` task would have done it if it could find this particular version of the SDK.  
   This is unfortunately still a very hacky approach and I will open up a follow-up PR that changes back to the saner pipeline later. For now, this works and is as good as can be expected with Preview bits.
* We should also use the .NET Core CLI tasks instead of custom scripts tasks, to get additional goodies from Azure Pipelines when it comes to parsing compiler error messages, etc.

# Description

## What Immo did wrong on Sunday's stream

1. Immo attempted to use the `UseDotNet@2` with the `release-version` listed as latest release on https://dotnet.microsoft.com/download/dotnet-core (i.e. `5.0.0-preview.3`).  
   The `UseDotNet@2` needs an SDK or Runtime version number (depending on the value for `packageType`, which he specified as `sdk`).  
   **The fix used in this PR**: Do exactly what Immo started with, but use the correct version number, or even better, don't care about exact version, and use `5.0.x` instead.
2. He then gave up on the `UseDotNet@2` and started using the now deprecated `DotNetCoreInstaller` task, that really should no longer be used, naturally using that old task lead to a couple of issues, including using wrong version, not getting the right channel, etc.

## Sorting out the various versions

There are multiple versions littered all across .NET Core.
* Channel version
   A list of .NET Core releases
* Release version  
   The version number listed in the table on https://dotnet.microsoft.com/download/dotnet-core
   This corresponds to the `release-version` property of an entry in https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/5.0/releases.json (replace the `5.0` to get another channel
* Runtime version
   Each .NET Core release has a separate Runtime version that is not equal to the *Release version*.
* SDK version
   Similar to Runtime version, each .NET Core Release also has a SDK version that is not equal to the *Release version*, nor to the *Runtime version*
* SDK version (Visual Studio style)
   If a .NET Core SDK is installed through the Visual Studio Installer, you get the same SDK as downloaded from https://dot.net/ **BUT the Visual Studio SDK package has an additional leading `0` in the patch version number!**

## What is `DotNetCoreInstaller` vs. `UseDotNet` in Azure Pipelines

`DotNetCoreInstaller` is the old version of `UseDotNet`. At one point, Azure Pipelines renamed the `DotNetCoreInstaller` task to `UseDotNet`. The hint is the version number of these tasks: 

* `DotNetCoreInstaller@0`
* `DotNetCoreInstaller@1`
* `UseDotNet@2`

**`DotNetCoreInstaller` is deprecated and has been superseded by `UseDotNet`.**

## Which inputs to use for `UseDotNet`

Ideally, you should let Microsoft figure out an appropriate version number. Use the wildcard versions with the trailing `.x`. So for .NET 5 use the following:

``` yml
- task: UseDotNet@2
  displayName: Use .NET 5 Preview
  inputs:
    version: 5.0.x
    packageType: sdk
    includePreviewVersions: true
``` 

But if you want to use a particular version number, be sure to use the exact version number on the actual download page of the particular release.  
That means: Open the actual channel page e.g. https://dotnet.microsoft.com/download/dotnet/5.0 and use the SDK version listed in the seconds column of the table.  
**Don't use the release version number listed on the general [.NET Core channel listing page](https://dotnet.microsoft.com/download/dotnet-core)**

## Why use `DotNetCoreCLI@2` instead of `script` tasks

Using `script` tasks to invoke the `dotnet` CLI on the command line works fine, but is limited in its functionality.

When you use the Azure Pipeline .NET Core CLI task `DotNetCoreCLI@2`, Azure Pipelines will actually add additional command-line arguments, like adding loggers. The custom logger that Pipelines adds, hooks directly into the compiler and intercepts warning and error messages emitted by the compiler. The advantage of this is: You get warnings and errors listed in the CI build summary nicely formatted instead of heaving to read through the raw text of the output log. Look at a [comparison picture](https://user-images.githubusercontent.com/8759693/80999294-b3949180-8e44-11ea-84c1-5a2d723e0661.png) to see the difference.

Also, when specifying the `test` command, the task will automatically determine how and where to emit the test result file and uploads it automatically. This is much nicer than having to do it manually yourself.

### CODE REVIEW

**Comment by ltrzesniewski** [comment]
Sorry for suggesting the `DotNetCoreInstaller` task - I had copied it from a build script of my own, I guess I should go update it now. 😉

**Comment by couven92** [comment]
Yeah, the rename from `DotNetCoreInstaller` to `UseDotNet` is not exactly obvious. Neither is how all these versioning schemes are connected.

**NOTE**

The "What did Immo do wrong" section is not about pointing out what @terrajobst did wrong! It's a good example of common mistakes anyone of us could easily make. This stuff is seriously hard! 

And take my explanations with a grain of salt, I only use a lot of time actively messing with my custom Azure Pipelines magics, I don't work for Microsoft and can only guess at what happened here, based on my own observations over time.

**Comment by couven92** [comment]
BTW, I don't know why the CI does not build this automatically, but for my own testing, I created my own CI pipeline for my fork. and I can see the CI run passing:

https://dev.azure.com/couven92/terrajobst-minsk/_build/results?buildId=429&view=results

**Comment by centreboard** [comment]
@couven92 If we wanted to pin the preview version (e.g. so breaking changes in source generation in later previews don't break the pipeline) am I correct we'd use `5.0.100-preview.3.20216.6` as the sdk version? This works for me just now, but didn't work for Immo last night.

> If a .NET Core SDK is installed through the Visual Studio Installer, you get the same SDK as downloaded from https://dot.net/

Are the 3.1 previews which ship with VS preview available to download with this task? 
They're not listed in https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/3.1/releases.json which only has the official previews from before 3.1.0.
Neither `3.1.300-preview-015135` nor `3.1.300-preview-15135` work for me.

**Comment by couven92** [comment]
@centreboard 

> If we wanted to pin the preview version (e.g. so breaking changes in source generation in later previews don't break the pipeline) am I correct we'd use 5.0.100-preview.3.20216.6 as the sdk version?

Correct, pinning using `5.0.100-preview.3.20216.6` is the correct version number and that worked for me as well. More precisely, you'd have to use a version number verbatim as listed in the `version` property of one of the `sdks` objects of an item listed in the `releases` array in the JSON document https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/5.0/releases.json  
i.e. `$.releases.*.sdks.*.version`

> This didn't work for Immo last night

I think it did actually, I think when he got that to work he was too quick to dismiss it, since the test run failed. Note how this PR also adds a task installing the .NET Core 3.1 runtime, since the test runner needs that to run. What happens here is that the `UseDotNet` task overrides the `dotnet` CLI, so you get a completely separate and clean `dotnet` install, the downside of this is that you have to explicitly install all versions with tasks youself, you cannot rely on a working version being present on the build agent any longer, i.e. with the extra control, you also get extra responsibility.

> Are the 3.1 previews which ship with VS preview available to download with this task?
They're not listed in https://dotnetcli.blob.core.windows.net/dotnet/release-metadata/3.1/releases.json which only has the official previews from before 3.1.0. Neither 3.1.300-preview-015135 nor 3.1.300-preview-15135 work for me.

Hmm, that is very interesting... But as you pointed out: A version that is not listed in the `releases.json` file for the appropriate channel (`3.1` in your example) cannot be installed by the task, since the task looks up the zip to download from this file (AFAIK). I think if a preview is superseded by a release version and there is no reason of keeping the original preview version, the preview is deleted.

**Comment by terrajobst** [comment]
> What Immo did wrong on Sunday's stream

 I 💖 this section. It's amazing to see a detailed report on where I went wrong and why. Thanks a lot for spending the time to write this up!

Side rant: Part of streaming programming is showcasing how developers deal with unexpected issues. We all stumble, we all make mistakes, and nobody is immune to that. Some people seem to think that people on the .NET product team are demigods but we aren't (well, at least I am not 😉). Also, as a member on the product team I know my own area extremely well, but that doesn't mean that I know the rest of the stack or the tooling well. There have been plenty of times where someone in the stream showed me a cool trick.

> When you use the Azure Pipeline .NET Core CLI task `DotNetCoreCLI@2`, Azure Pipelines will actually add additional command-line arguments, like adding loggers. The custom logger that Pipelines adds, hooks directly into the compiler and intercepts warning and error messages emitted by the compiler. The advantage of this is: You get warnings and errors listed in the CI build summary nicely formatted instead of heaving to read through the raw text of the output log.

💜💜💜

> Also when specifying the `test` command, the task will automatically determine how and where to emit the test result file and uploads it automatically. This is much nicer than having to do it manually yourself.

🥰

> * You should use the `UseDotNet@2` task (not the deprecated `DotNetCoreInstaller` task)
> * Instead of using a particular version, use the `5.0.x` wildcard version and enable preview builds
> * We should also use the .NET Core CLI tasks instead of custom scripts tasks, to get additional goodies from Azure Pipelines when it comes to parsing compiler error messages, etc.

Those all make sense to me. However, at least locally it seems that using the .NET 5 SDK results in our generator to not run at all. I suspect the reason it works for you is because I committed the generated file.

**Comment by terrajobst** [comment]
> Those all make sense to me. However, at least locally it seems that using the .NET 5 SDK results in our generator to not run at all. I suspect the reason it works for you is because I committed the generated file.

#118 [confirms](https://terrajobst.visualstudio.com/Minsk/_build/results?buildId=1247&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=2b20c7d0-7587-5b0f-aadc-b456e6f0b807) my suspicion :-(

**Comment by couven92** [comment]
> However, at least locally it seems that using the .NET 5 SDK results in our generator to not run at all. I suspect the reason it works for you is because I committed the generated file.

@terrajobst Oh, I missed that of course... Obviously I'm not a demigod either! 😛 I'll investigate and see if I manage to find more insights until Wednesday.

**Comment by couven92** [comment]
@terrajobst Just for illustration, but look at the differences of the build summary after my change that uses the `DotNetCoreCLI@2` task:

azure-pipelines-build-summary-comparison
*(left side using `script` tasks, right side using `DotNetCoreCLI@2`)*

File: azure-pipelines.yml [hunk 1]
```diff
@@ -12,13 +12,28 @@ variables:
...
+  displayName: Use .NET 5.0 Preview
+  inputs:
+    version: 5.0.x
+    packageType: sdk
+    includePreviewVersions: true
+# We now still need .NET Core 3.1 Runtime for CI tests
+- task: UseDotNet@2
+  displayName: Install .NET Core 3.1 Runtime
+  inputs:
+    version: 3.1.x
```
Comment (cartermp) [inline azure-pipelines.yml:25]: Another option is to specify your `global.json` with `rollForward` set and build based on that like so:https://github.com/fsprojects/FsLexYacc/blob/master/azure-pipelines.yml#L17-L22

Comment (couven92) [reply azure-pipelines.yml:25]: @cartermp I guess the advantage of that is that you'd be using the same settings for both local development and CI builds, without having to specify/maintaining the versions in multiple places, correct? Using this, `global.json` becomes your one source of truth (almost-ish)?

Comment (cartermp) [reply azure-pipelines.yml:25]: Yep! I personally always use global.json for anything other than toy projects

Comment (couven92) [reply azure-pipelines.yml:25]: @cartermp But the rolling forward does not work very well with Preview versions unfortunenately, so there you're stuck with pinning a specific version anyways. And you still need to specify the base version-wildcard in both the Azure Pipelines YAML and the global.json if I read your link correctly?

**Comment by couven92** [comment]
@terrajobst Phew! Okay, with commit https://github.com/terrajobst/minsk/pull/116/commits/0b746083d39cb59b07413f10588e37d869985573 I finally got everything working.

I noticed that the [dotnet-install scripts](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-install-script) were able to download the preview-sdk version you had installed locally. So I temporarily commented our the `UseDotNet@2` task and replaced it with a bash-script that emulates the task as best it can. **This is a temporary fix** and I added comments indicating that we should change back to the `UseDotNet@2` task as soon as a working Preview SDK is properly listed in the releases list.

https://dev.azure.com/couven92/terrajobst-minsk/_build/results?buildId=437&view=results

**Comment by centreboard** [comment]
@couven92 
Awesome.

I found an alternative workaround is to temporarily add a package reference to [Microsoft.Net.Compilers.Toolset](https://www.nuget.org/packages/Microsoft.Net.Compilers.Toolset/3.6.0-3.final) into the Minsk project to force it to use a newer version of the compiler.
https://github.com/dotnet/roslyn-sdk/issues/524#issuecomment-623652809

An advantage is it would enable anyone running the .Net 5 preview to be able to build locally without needing to add their own global.json, but we'd need to remember to remove it or keep it updated.

**Comment by Nirmal4G** [comment]
This should go into the azure pipelines documentation. Very well explained.

File: azure-pipelines.yml [hunk 2]
```diff
@@ -12,18 +12,26 @@ variables:
...
-    version: 3.1.x
-    packageType: runtime
+# TODO: Remove the following crazy hack once a properly listed Preview SDK for Source Generators feature becomes available.
+- script: |
+    curl -o "$AGENT_TEMPDIRECTORY/dotnet-install.sh" -sSL https://dot.net/v1/dotnet-install.sh
+    bash "$AGENT_TEMPDIRECTORY/dotnet-install.sh" --version 3.1.300-preview-015135 --install-dir "$AGENT_TOOLSDIRECTORY/dotnet" --verbose
+    echo "##vso[task.setvariable variable=Dotnet.Root;]$AGENT_TOOLSDIRECTORY/dotnet"
+    echo "##vso[task.prependpath]$AGENT_TOOLSDIRECTORY/dotnet"
+  displayName: Use .NET Core 3.1.3 Preview SDK
+# TODO: Uncomment lines below once Preview/Release SDK for Source Generators feature becomes available.
```
Comment (Nirmal4G) [inline azure-pipelines.yml:22]: We have `git revert` so, it's best that we don't include commented out section in the commit.

File: azure-pipelines.yml (hunk 1 above)
Comment (cartermp) [reply azure-pipelines.yml]: It _should_ work with the preview setting specified, but if that doesn't then I'd consider that a bug I suppose

File: azure-pipelines.yml (hunk 2 above)
Comment (terrajobst) [reply azure-pipelines.yml:22]: Yeah, but in this case I prefer the commented version because it provides additional context.
